### PR TITLE
refactor(routing): standardize cost terminology

### PR DESCRIPTION
**Pull request by Juan-Francisco-Robles**
# Description

## Summary

This PR standardizes naming across the routing package by replacing
`distance` terminology with `cost` terminology wherever the value
represents the generic cuOpt cost matrix rather than a physical distance.

The change improves semantic consistency across routing internals and
aligns the implementation with the existing `data_model_view.hpp`
documentation, which already describes the cost matrix as supporting
arbitrary user-defined metrics (e.g. time, miles, meters, euros, or
other real-valued costs).

All changes are limited to the `routing` package and do not modify
optimization behavior.

## Changes Made

* Renamed `distance`-related routing helpers, variables, and types to
  `cost`-based equivalents where appropriate.
* Replaced:

  * `distance_dim` → `cost_dim`
  * `distance_node_t` → `cost_node_t`
  * `distance_route_t` → `cost_route_t`
* Renamed:

  * `distance_node.cuh` → `cost_node.cuh`
  * `distance_route.cuh` → `cost_route.cuh`
* Updated all dependent routing internals and local-search consumers.
* Adjusted wording in `data_model_view.hpp` to consistently refer to
  generic optimization costs rather than physical distance.

## Scope

Affected components include:

* routing node and route helpers
* local search internals
* GES lexicographic search
* diversity helpers/macros
* routing problem internals
* utility kernels

## Impact

* No behavioral or algorithmic changes
* No optimization-result changes
* Purely semantic/cosmetic refactor
* Improves clarity and maintainability
* Better separates generic optimization costs from physical distance concepts

## Validation

* Project builds successfully
* Existing routing tests pass

## Notes

This change touches internal routing identifiers and file names.
Downstream branches using the old `distance_*` internal symbols may
require rebasing or reference updates.

## Issue

N/A

## Checklist

- [x] I am familiar with the Contributing Guidelines.
- Testing
   - [x] New or existing tests cover these changes
   - [ ] Added tests
   - [ ] Created an issue to follow-up
   - [x] NA
- Documentation
   - [x] The documentation is up to date with these changes
   - [ ] Added new documentation
   - [x] NA

### CODE REVIEW

File: cpp/src/routing/route/dimensions_route.cuh [hunk 1]
```diff
@@ -265,7 +265,7 @@ class dimensions_route_t {
   request_route_t<i_t, f_t, REQUEST> requests;
 
   // distance route
```
Comment (rg20) [inline cpp/src/routing/route/dimensions_route.cuh]:
```suggestion
  // cost route
```

File: cpp/src/routing/route/mismatch_route.cuh [hunk 2]
```diff
@@ -138,7 +138,7 @@ class mismatch_route_t {
   }
 
   /**
-   * @brief Get the shared memory size required to store a distance route of a given size
+   * @brief Get the shared memory size required to store a cost route of a given size
```
Comment (rg20) [inline cpp/src/routing/route/mismatch_route.cuh]:
```suggestion
   * @brief Get the shared memory size required to store a mismatch route of a given size
```

File: cpp/src/routing/route/prize_route.cuh [hunk 3]
```diff
@@ -148,7 +148,7 @@ class prize_route_t {
   }
 
   /**
-   * @brief Get the shared memory size required to store a distance route of a given size
+   * @brief Get the shared memory size required to store a cost route of a given size
```
Comment (rg20) [inline cpp/src/routing/route/prize_route.cuh]:
```suggestion
   * @brief Get the shared memory size required to store a prize route of a given size
```

File: cpp/src/routing/route/service_time_route.cuh [hunk 4]
```diff
@@ -140,7 +140,7 @@ class service_time_route_t {
   }
 
   /**
-   * @brief Get the shared memory size required to store a distance route of a given size
+   * @brief Get the shared memory size required to store a cost route of a given size
```
Comment (rg20) [inline cpp/src/routing/route/service_time_route.cuh]:
```suggestion
   * @brief Get the shared memory size required to store a service time route of a given size
```

File: cpp/src/routing/route/tasks_route.cuh [hunk 5]
```diff
@@ -135,7 +135,7 @@ class tasks_route_t {
   }
 
   /**
-   * @brief Get the shared memory size required to store a distance route of a given size
+   * @brief Get the shared memory size required to store a cost route of a given size
```
Comment (rg20) [inline cpp/src/routing/route/tasks_route.cuh]:
```suggestion
   * @brief Get the shared memory size required to store a tasks route of a given size
```

File: cpp/src/routing/route/vehicle_fixed_cost_route.cuh [hunk 6]
```diff
@@ -104,7 +104,7 @@ class vehicle_fixed_cost_route_t {
   }
 
   /**
-   * @brief Get the shared memory size required to store a distance route of a given size
+   * @brief Get the shared memory size required to store a cost route of a given size
```
Comment (rg20) [inline cpp/src/routing/route/vehicle_fixed_cost_route.cuh]:
```suggestion
   * @brief Get the shared memory size required to store a vehicle fixed cost route of a given size
```

**Review by rg20 — COMMENTED** [review]
Thanks for the PR. 

I have minor comments, these already caught by coderabbit

File: cpp/src/routing/route/dimensions_route.cuh [hunk 7]
```diff
@@ -265,7 +265,7 @@ class dimensions_route_t {
...
-  distance_route_t<i_t, f_t> distance_dim;
+  cost_route_t<i_t, f_t> cost_dim;
```
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/dimensions_route.cuh]: Thank you. This issue has been addressed in subsequent issues.

File: cpp/src/routing/route/mismatch_route.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/mismatch_route.cuh]: Thank you. This issue has been resolved commit: 
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/prize_route.cuh (hunk 3 above)
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/prize_route.cuh]: This issue has been resolved commit: 
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/service_time_route.cuh (hunk 4 above)
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/service_time_route.cuh]: Issue resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/tasks_route.cuh (hunk 5 above)
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/tasks_route.cuh]: Issue resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/vehicle_fixed_cost_route.cuh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/vehicle_fixed_cost_route.cuh]: Issue resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/util_kernels/set_nodes_data.cuh [hunk 8]
```diff
@@ -54,8 +54,8 @@ __device__ void set_route_data(typename problem_t<i_t, f_t>::view_t const& probl
                        route.template get_dim<dim_t::TIME>().excess_backward[0]) < 0.0001,
                    "Backward forward mismatch!");
     }
-    route.template get_dim<dim_t::DIST>().distance_backward[n_nodes_route] = 0.f;
-    route.template get_dim<dim_t::DIST>().distance_forward[0]              = 0.f;
+    route.template get_dim<dim_t::COST>().cost_backward[n_nodes_route] = 0.f;
+    route.template get_dim<dim_t::COST>().cost_forward[0]              = 0.f;
```
[the remark this answers is not in the document]
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/util_kernels/set_nodes_data.cuh:58]: Thank you for the suggestion. I agree that this route boundary initialization path should be covered by a regression test after the DIST → COST semantic standardization.

I’ve added a targeted routing test covering both VRP and PDP route setup paths by dirtying the COST boundary buffers and verifying that `set_route_data` correctly resets `cost_forward[0]` and `cost_backward[n]`.

Related commit:
test(routing): add routing test for COST boundary initialization

File: cpp/src/routing/route/dimensions_route.cuh (hunk 1 above)
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/dimensions_route.cuh]: Thank you @rg20 for pointing this out. This modification was not correctly applied during the standardization process carried out in this PR. Issue resolved in commit: 
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/mismatch_route.cuh (hunk 2 above)
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/mismatch_route.cuh]: As done in the previous issue, this one has been also resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/prize_route.cuh (hunk 3 above)
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/prize_route.cuh]: Resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/service_time_route.cuh (hunk 4 above)
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/service_time_route.cuh]: Resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/tasks_route.cuh (hunk 5 above)
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/tasks_route.cuh]: Resolved in commit:
refactor(routing): update distance reference comments in routing files.

File: cpp/src/routing/route/vehicle_fixed_cost_route.cuh (hunk 6 above)
Comment (Juan-Francisco-Robles) [reply cpp/src/routing/route/vehicle_fixed_cost_route.cuh]: Resolved in commit:
refactor(routing): update distance reference comments in routing files.

**Comment by Juan-Francisco-Robles** [comment]
All review comments have been addressed.
Added a routing test covering COST boundary initialization to ensure `set_route_data` correctly resets `cost_forward[0]` and `cost_backward[n]`.
Ready for re-review.

**Comment by tmckayus** [comment]
@josembd please see the error from check_style below. If you run the pre-commit hooks to fix the style errors, then the cpp build will run and we can get full test results.
--
If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
To run `pre-commit` as part of git workflow, use `pre-commit install`.

**Comment by josembd** [comment]
Thanks for pointing this out, @tmckayus. The routing pre-commit issues have been fixed in 9f309421, and the relevant checks pass locally.

Could you please authorize CI for this SHA with `/ok to test 9f309421`?

**Review by rg20 — APPROVED** [review]

**Comment by josembd** [comment]
Hi @rg20, thanks for reviewing and approving the PR.

We noticed that this PR is still open in the 26.08 milestone, which is due today. Could you let us know whether anything else is needed from our side before it can be merged, such as updating the branch with the latest changes, rerunning the checks, or retargeting the PR to the `release/26.08` branch?

We also have another contribution that introduces a separate feature, but its implementation depends on the changes introduced in this PR. More generally, for this and future contributions with technical dependencies between them, what workflow do you prefer? Should we wait until the prerequisite PR is merged before opening the dependent one, or is it acceptable to open it as a stacked PR in the meantime?

Since these are among our first contributions to cuOpt, we want to make sure we are following the workflow that is most convenient for the maintainers.

Thanks for your guidance.

**Comment by ramakrishnap-nv** [comment]
> Hi @rg20, thanks for reviewing and approving the PR.
> 
> We noticed that this PR is still open in the 26.08 milestone, which is due today. Could you let us know whether anything else is needed from our side before it can be merged, such as updating the branch with the latest changes, rerunning the checks, or retargeting the PR to the `release/26.08` branch?
> 
> We also have another contribution that introduces a separate feature, but its implementation depends on the changes introduced in this PR. More generally, for this and future contributions with technical dependencies between them, what workflow do you prefer? Should we wait until the prerequisite PR is merged before opening the dependent one, or is it acceptable to open it as a stacked PR in the meantime?
> 
> Since these are among our first contributions to cuOpt, we want to make sure we are following the workflow that is most convenient for the maintainers.
> 
> Thanks for your guidance.

Hi @Juan-Francisco-Robles, 

Since the release is today, we normally don't merge any new PRs since there are chances of breakages. So I will move this to 26.10 version.

**Comment by josembd** [comment]
Hi @rg20 @ramakrishnap-nv, just following up on this PR now that we're moving through the 26.10 cycle.

The PR was previously approved and moved from the 26.08 milestone to 26.10, and we've kept the branch updated with `main` since then.

Is there anything else needed from our side to get this ready for merge? We also have a follow-up contribution that depends on these changes, so it would be useful for us to know whether we should continue rebasing on top of this PR or if you expect #1299 to be merged soon.

Happy to update the branch, rerun CI, or address any additional review comments if needed. Thanks!

**Review by ramakrishnap-nv — APPROVED** [review]

**Comment by ramakrishnap-nv** [comment]
Thank you @josembd, I have approved, will wait for the CI to complete.

**Comment by ramakrishnap-nv** [comment]
`checks / check-style` is failing on the `verify-copyright` pre-commit hook — the SPDX header in 4 files touched by this PR is still `...-2025` and needs bumping to `...-2026`. I don't have push access to this branch (fork PR), so here's the exact patch — apply with `git apply` (or update the 4 lines by hand) and push:

```diff
diff --git a/cpp/src/routing/node/break_node.cuh b/cpp/src/routing/node/break_node.cuh
index d56596e8..48a49d86 100644
--- a/cpp/src/routing/node/break_node.cuh
+++ b/cpp/src/routing/node/break_node.cuh
@@ -1,6 +1,6 @@
 /* clang-format off */
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 /* clang-format on */
diff --git a/cpp/src/routing/node/mismatch_node.cuh b/cpp/src/routing/node/mismatch_node.cuh
index d32a2224..c4fa59c1 100644
--- a/cpp/src/routing/node/mismatch_node.cuh
+++ b/cpp/src/routing/node/mismatch_node.cuh
@@ -1,6 +1,6 @@
 /* clang-format off */
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2023-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2023-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 /* clang-format on */
diff --git a/cpp/src/routing/node/service_time_node.cuh b/cpp/src/routing/node/service_time_node.cuh
index 8bcf3da8..d7133c0a 100644
--- a/cpp/src/routing/node/service_time_node.cuh
+++ b/cpp/src/routing/node/service_time_node.cuh
@@ -1,6 +1,6 @@
 /* clang-format off */
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2023-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2023-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 /* clang-format on */
diff --git a/cpp/src/routing/node/tasks_node.cuh b/cpp/src/routing/node/tasks_node.cuh
index dd985533..1ffb72a0 100644
--- a/cpp/src/routing/node/tasks_node.cuh
+++ b/cpp/src/routing/node/tasks_node.cuh
@@ -1,6 +1,6 @@
 /* clang-format off */
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2023-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2023-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 /* clang-format on */
```

Verified this applies cleanly against the current branch tip (`3fe207d2`). Enabling "Allow edits by maintainers" on this PR would let a maintainer push straightforward fixes like this directly next time.

**Comment by josembd** [comment]
Hi @ramakrishnap-nv,

Thank you for your response and for pointing this out. I have applied the requested copyright updates and pushed them in commit 6b9a70ab.

I am a member of the organization that owns the fork, so I can push changes to the branch. However, GitHub does not provide the “Allow edits from maintainers” option for organization-owned forks.

Please let me know if any further changes are needed and I will apply them promptly. Could you please rerun the CI checks when convenient? Thanks again for the review and guidance.

**Comment by ramakrishnap-nv** [comment]
## `wheel-build-libcuopt` failure: branch is out of date, not a bug in this PR

The build fails at:
```
cpp/src/routing/utilities/md_utils.hpp:174:6: error: variable 'stream' set but not used [-Werror=unused-but-set-variable]
```

This is a known build break that was already fixed on `main` in #1894 (merged as `3da6c366`, "fix(build): repair cuda::stream_ref migration breakage from #1828"). It was caused by #1828's stream-API migration leaving a couple of call sites broken — unrelated to anything in this PR's diff.

This branch is currently **205 commits behind** `NVIDIA/cuopt:main`, from before that fix landed, which is why the build still fails here.

**Fix:** merge/rebase this branch onto the latest `NVIDIA/cuopt:main` to pick up #1894, then re-push. No changes needed in this PR's own commits.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by ramakrishnap-nv** [comment]
**Correction:** my "205 commits behind" figure above was wrong — I'd compared against the fork's own stale `main` (`ogaai/cuopt:main`), not `NVIDIA/cuopt:main`. Using GitHub's actual computed merge-base for this PR (`67773af5`), this branch is only **13 commits behind** `NVIDIA/cuopt:main`, not 205. Sorry for the confusion.

The underlying diagnosis still holds: `3da6c366` (#1894, the `md_utils.hpp` fix) is one of those 13 commits, so a rebase/merge onto current `main` should still resolve the build failure.

**Comment by josembd** [comment]
Thanks for investigating and identifying the cause, @ramakrishnap-nv. I have updated the branch with the latest NVIDIA/cuopt:main. The new head SHA is 463f686 and now includes the fix from #1894.

The branch should be ready for CI again. Thanks for your help.

**Comment by josembd** [comment]
@ramakrishnap-nv The updated CI has one remaining failure: INCUMBENT_CALLBACK_TEST aborted at the same current_incumbent.size() == worker->leaf_problem.num_cols assertion tracked in #1909. All 42 files changed by this PR are under routing; it does not modify branch_and_bound.cpp or the incumbent callback tests, so this appears to be the known unrelated intermittent race. The other 1030 tests passed. Could you advise whether the failed job should be rerun or treated as a known failure?

**Comment by ramakrishnap-nv** [comment]
@josembd It is a flaky test and we are working on it. I have retriggered the PR.
